File: ios/Keyboard/Layout.swift
import Foundation

/// The canonical character table, same one as `build_keyman.py` and
/// `windows/src/KBDMSSTD.klc`. Standard orthography only (HOUSE_STYLE §1):
/// the extended alphabet is deliberately absent everywhere, not just here.
///
/// A phone has no AltGr, so the four letters live on LONGPRESS of their base
/// key - `c` held gives `č`. That is the same mnemonic as AltGr+C on desktop,
/// which is the point: one habit across platforms.
enum Layout {

    /// base letter -> the accented letter reached by holding it
    static let accents: [Character: Character] = [
        "c": "č",
        "s": "š",
        "z": "ž",
        "e": "ě",
    ]

    /// Punctuation our MS texts actually use, held on the `.` key.
    static let periodAccents: [Character] = [",", "!", "?", "„", "”", "'", ":", ";"]

    /// Dashes, held on `-` in the numeric layer.
    static let hyphenAccents: [Character] = ["–", "—"]

    static let letterRows: [[Character]] = [
        Array("qwertyuiop"),
        Array("asdfghjkl"),
        Array("zxcvbnm"),
    ]

    static let numericRows: [[Character]] = [
        Array("1234567890"),
        Array("@#$%&*()"),
        Array("-/:;'\"?!"),
    ]

    /// Longpress variants for any key, or nil if the key has none.
    static func variants(for key: Character, uppercase: Bool) -> [Character]? {
        if let accent = accents[Character(key.lowercased())] {
            let letter = uppercase ? Character(accent.uppercased()) : accent
            return [letter]
        }
        if key == "." { return periodAccents }
        if key == "-" { return hyphenAccents }
        return nil
    }
}

File: android-app/app/src/main/kotlin/com/radoslove/interslavic/Layout.kt
package com.radoslove.interslavic

/**
 * The canonical character table — the same one as `docs/ms-latin-table.md`,
 * `ios/Keyboard/Layout.swift`, `build_keyman.py` and the Android layout XML.
 * Standard orthography only (HOUSE_STYLE §1): the extended alphabet
 * (ų ė ę ȯ å ŕ ť ď ľ đ …) is deliberately absent everywhere, not just here.
 *
 * A phone has no AltGr, so the four letters live on LONGPRESS of their base key
 * — hold `c` to get `č`, the same mnemonic as AltGr+C on desktop. The three
 * digraphs sit on longpress of `d` / `l` / `n`, matching `android/isv_latin.xml`.
 *
 * If that table ever changes, this file is one more place that changes with it.
 */
object Layout {

    /** base letter -> the accented letter reached by holding it */
    val accents: Map<Char, Char> = mapOf(
        'c' to 'č',
        's' to 'š',
        'z' to 'ž',
        'e' to 'ě',
    )

    /** base letter -> the digraph reached by holding it */
    val digraphs: Map<Char, String> = mapOf(
        'd' to "dž",
        'l' to "lj",
        'n' to "nj",
    )

    val letterRows: List<String> = listOf(
        "qwertyuiop",
        "asdfghjkl",
        "zxcvbnm",
    )

    /**
     * Symbol / numeric layer (the ?123 page). Row 3 leads with the canonical MS
     * punctuation as direct keys — „ ” – — and ’ — so the four glyphs from
     * `docs/ms-latin-table.md` are reachable without a popup picker. `?` and the
     * rest of the common punctuation live here too; on the letter layer only
     * `.` and `,` are exposed directly.
     */
    val symbolRows: List<String> = listOf(
        "1234567890",
        "@#\$%&*()/-",
        "„”–—’:;!?",
    )

    /**
     * Longpress output for a base key given the shift state, or null if the key
     * has no longpress. Uppercase is applied explicitly here — the iOS caps bug
     * was a reminder that the shifted accent must be produced deliberately, not
     * assumed.
     */
    fun longPress(key: Char, uppercase: Boolean): String? {
        accents[key]?.let { acc ->
            return if (uppercase) acc.uppercaseChar().toString() else acc.toString()
        }
        digraphs[key]?.let { dg ->
            return if (uppercase) dg.replaceFirstChar { it.uppercaseChar() } else dg
        }
        return null
    }
}
